### DISCUSSION TITLE
Slack notification channel

### DISCUSSION THREAD

**Discussion by alexjustesen** [alexjustesen/speedtest-tracker#1220] [Q&A]
@jonezy35 extending our conversation from https://github.com/alexjustesen/speedtest-tracker/pull/1196#issuecomment-1959590832...

I'm about to merge a PR that moves the settings migrations to the correct location #1219. When you go to create the Slack settings use the command `sail artisan make:settings-migration CreateSlackNotificationSettings` and this generate a file in the correct location for you.

Regarding using Laravel's `Http` client, I think I prefer using Spatie's package for sending `WebhookCall`, you can check out how that's used in https://github.com/alexjustesen/speedtest-tracker/blob/main/app/Actions/Notifications/SendWebhookTestNotification.php and their docs are located here https://github.com/spatie/laravel-webhook-server.

I'm going to refactor the Discord test to use `WebhookCall` instead, if you can hold off on Slack until I merge that change (later this morning) you can leverage it for Slack.

**Comment by jonezy35** [comment]
@alexjustesen sounds good! I used `http` as I was having trouble getting `WebhookCall` to work due to the way I formatted the payload so I'm interested to see what the final code looks like. Discord only expects certain data formats, `content` being the main one. I've got a busy few days of work so likely won't get started on Slack until tomorrow at the earliest.

**Reply by alexjustesen** [reply]
Sounds good, here's the PR with some of my changes. I missed a couple of things in the initial PR review so made some adjustments.

https://github.com/alexjustesen/speedtest-tracker/pull/1221

**Reply by alexjustesen** [reply]
Oh important note, using WebhookCall allows us to put the request into the queue for processing so it can be processed in the background by a queue worker.

You _can_ do that with actions as well but this helps separate code that should run synchronously vs. asynchronously.

**Comment by jonezy35** [comment]
Regarding the docker image question, how long does it take for new features to be present in the image, and what is the process for the image to be updated with the new features when a merge is made?

**Reply by alexjustesen** [reply]
Process is...

1. I create a release branch which bumps the build information, merge that into `main` as it's considered "production code".
2. I tag a release which generates the release notes and builds the compressed assets.
3. LSIO has a build process that runs hourly that checks for any tagged releases, if there is a new one they build a new image.

**Reply by jonezy35** [reply]
That's awesome! More from a curiosity perspective on my part, how did you work with LSIO to get your version of `speedtest-tracker` to be served up by them? I'm currently very uninformed as to LSIO's process for what containers they serve and how those containers are maintained.

**Reply by alexjustesen** [reply]
I posted that I needed help with building an image and probably through repo popularity one of the guys saw it: https://github.com/alexjustesen/speedtest-tracker/issues/837

No idea what the real process is but I'd assume it's documented somewhere on their Discord.

**Comment by jonezy35** [comment]
@alexjustesen I see laravel's `http` client is still in the listeners for Discord. Did you decide to keep it or were you meaning you prefer `WebhookCall` moving forward for the Slack notifications?

**Reply by alexjustesen** [reply]
Nah, got distracted and forgot. If you want to do a PR to bring that into alignment that be cool.

**Reply by jonezy35** [reply]
@alexjustesen are test discord notifications working for you with `WebhookCall`? I pulled the most recent code to start working on this and the test notifications aren't working for me. (it says the notification was sent but Discord never receives it). But when I switch it back to laravel's `http` it works.

I don't have a ton of time today so I wanted to check with you before I troubleshoot furthur. If it's not working for you either I can open an Issue to track it and I can get to it later today

**Reply by jonezy35** [reply]
It also looks like there's a [Laravel Discord alerts](https://github.com/spatie/laravel-discord-alerts) as well as a [Laravel Slack alerts](https://github.com/spatie/laravel-slack-alerts)

**Reply by jonezy35** [reply]
Nevermind... I forgot `WebhookCall` adds the job to the queue so I had to run `./vendor/bin/sail artisan queue:work` to get it to send. I'll go through and update the listeners to use `WebhookCall`.

For Slack would you be open to using [laravel-slack-alerts](https://github.com/spatie/laravel-slack-alerts)?

**Reply by alexjustesen** [reply]
So if I remember correctly Slack has been trying to get rid of Webhooks on their platform which is what the Spatie package uses. I think it might be better to go with the 1st party option and create a Slack App. I believe they went the Slack app route because it requires the workspace admin to approve it's access. From a data security standpoint that's a better practice than letting "anyone" send data/messages to the channel.

- Package: https://laravel.com/docs/10.x/notifications#slack-notifications
- Creating a Slack App: https://api.slack.com/apps?new_app=1

**Reply by jonezy35** [reply]
Makes total sense. Thank you for the documentation!

**Comment by jonezy35** [comment]
> Nah, got distracted and forgot. If you want to do a PR to bring that into alignment that be cool.

Submitted in PR #1242
